src/app_engine/gcmrecord.py

```diff
@@ -61,94 +161,118 @@ def get_associated_records_for_gcm_id(cls, gcm_id, verified_only = False):
+                         code=code,
+                         verified=False,
+                         code_sent_time=now,
+                         last_modified_time=now)
+      encrypted_record = GCMEncryptedRecord(
+          parent=get_ancestor_key(),
+          user_key=user_key,
+          gcm_key=gcm_key,
+          content=record.get_encrypted())
+      encrypted_record.put()
+      logging.info('GCM binding added, user_id=%s, gcm_id=%s', user_id, gcm_id)
       return constants.RESPONSE_CODE_SENT
 
-    assert len(records) == 1
-    if records[0].verified:
-      logging.warning('Cannot update GCM binding code since already verified, '\
-          'user_id=%s, gcm_id=%s' % (user_id, gcm_id))
+    assert len(encrypted_records) == 1
+    encrypted_record = encrypted_records[0]
+    record = GCMRecord.parse_encrypted(encrypted_record.content)
+    if not record:
+      encrypted_record.key.delete()
+      return constants.RESPONSE_INTERNAL_ERROR
+    if record.verified:
+      logging.warning('Cannot update GCM binding code since already verified, '
+                      'user_id=%s, gcm_id=%s', user_id, gcm_id)
       return constants.RESPONSE_INVALID_STATE
 
-    records[0].code = code
-    records[0].code_sent_time = now
-    records[0].last_modified_time = now
-    records[0].put()
+    record.code = code
+    record.code_sent_time = now
+    record.last_modified_time = now
+    encrypted_record.content = record.get_encrypted()
+    encrypted_record.put()
     logging.info(
-          'GCM binding code updated, user_id=%s, gcm_id=%s' % (user_id, gcm_id))
+        'GCM binding code updated, user_id=%s, gcm_id=%s', user_id, gcm_id)
     return constants.RESPONSE_CODE_RESENT
 
   @classmethod
   @ndb.transactional(retries=100)
   def verify(cls, user_id, gcm_id, code):
-    records = GCMRecord.get_by_ids(user_id, gcm_id)
-    assert len(records) < 2
-    for record in records:
-      if record.verified:
-        logging.warning('GCM binding already verified, user_id=%s, gcm_id=%s' \
-            % (user_id, gcm_id))
-        return constants.RESPONSE_INVALID_STATE
-
-      if record.code == code:
-        record.verified = True;
-        record.last_modified_time = datetime.datetime.utcnow()
-        record.put()
-        logging.info(
-            'GCM binding verified, user_id=%s, gcm_id=%s' % (user_id, gcm_id))
-        return constants.RESPONSE_SUCCESS
-      else:
-        return constants.RESPONSE_INVALID_CODE
-    return constants.RESPONSE_NOT_FOUND
+    """Marks the registration as verified if the supplied code matches."""
+    gcm_key = Hasher.salted_hash(gcm_id)
+    encrypted_records = GCMEncryptedRecord.get_by_gcm_key(gcm_key)
+    assert len(encrypted_records) < 2
+    if not encrypted_records:
+      return constants.RESPONSE_NOT_FOUND
+    encrypted_record = encrypted_records[0]
+    record = GCMRecord.parse_encrypted(encrypted_record.content)
+    if not record:
+      encrypted_record.key.delete()
```

Comment:
On second thought, let's not delete so we can potentially see what happened by examining the store.